hs_core/search_indexes.py

```diff
@@ -100,6 +100,56 @@ def normalize_name(name):
     return normalized.strip()
 
 
+# What extensions indicate each kind of high-level content type
+extension_content_types = {
+    'Document': set(['doc', 'docx', 'pdf', 'odt', 'rtf', 'tex', 'latex']),
+    'Spreadsheet': set(['csv', 'xls', 'xlsx', 'ods']),
+    'Presentation': set(['ppt', 'pptx', 'odp']),
+    'Jupyter Notebook': set(['ipynb']),
+    'Image': set(['gif', 'jpg', 'jpeg', 'tif', 'tiff', 'png']),
+    'Multidimensional (NetCDF)': set(['nc']),
+    'Multidimensional (HDF)': set(['hdf']),
+}
+
+
+def get_content_types(res):
+    """ return a set of content types matching extensions in a resource.
+        These include content types of logical files, as well as the generic
+        content types 'Document', 'Spreadsheet', 'Presentation'.
+
+        This is only meaningful for Generic or Composite resources.
+    """
+
+    resource = res.get_content_model()  # enable full logical file interface
+
+    types = set([res.discovery_content_type])  # accumulate high-level content types.
+    exts = set()  # track individual file extensions
+
+    # categorize logical files by type, and files without a logical file by extension.
+    for f in resource.files.all():
+        if f.has_logical_file:
+            candidate_type = type(f.logical_file).get_discovery_content_type()
+            types.add(candidate_type)
+        else:  # collect extensions of files not corresponding to logical metadata
```

Comment:
Is this really an either/or situation?  Why not categorize content type by both logical_file as well as extensions?

Reply:
I actually do. Here is the logic: 
 
* Anything with an aggregation is already completely documented beyond doubt. An aggregation is a heavyweight piece of metadata. 
* It is presumed that anything with an aggregation is characterized completely by the aggregation. This is hard evidence of what the resource entails. 
* By contrast, file extensions are a soft determination. People can name something .nc and not mean it! 
* So, I prefer to utilize the hard determination as definitive, and only use extensions when I must. 
* Thus the algorithm marks all files documented by aggregations as accounted for, and then only marks the files that are left over by extension. 
* Alas, there are some files that should have aggregations and don't. These include: 

a. files in generic resources. 
b. files whose aggregations were deleted by a user for whatever reason. 
c. files arising from a time when aggregations were optional. 

etc. 

I have every belief that eventually, all of these will have their own aggregations and this mechanism will be redundant. Thus I have designed in its gradual retirement as aggregations are created.